packages/wdio-utils/src/driver/utils.ts

```diff
@@ -125,7 +210,289 @@ export async function setupChrome(cacheDir: string, caps: Capabilities.Capabilit
+        }[OS],
+
+        get build() {
+            return [dlURL.baseUrl, dlURL.path, dlURL.filename].join('/') + dlURL.ext
+        }
+    }
+
+    const url = dlURL.build
+
+    log.info(`Setting up Firefox v${mozVersion}`)
+
+    const fileName: string = url.toString().split('/').pop() ?? ''
+
+    const cache = new Cache(cacheDir)
+    const browserRoot = cache.browserRoot(Browser.FIREFOX)
+    const archivePath = path.join(
+        browserRoot,
+        `${mozVersion}${path.extname(fileName)}`
+    )
+    if (!fs.existsSync(browserRoot)) {
+        fs.mkdirSync(browserRoot, { recursive: true })
+    }
+
+    const outputPath = cache.installationDir(
+        Browser.FIREFOX,
+        arch as BrowserPlatform,
+        channel === 'FIREFOX_DEVEDITION'
+            ? `${mozVersion}dev`
+            : mozVersion
+    )
+    fs.mkdirSync(outputPath, { recursive: true })
+
+    const getExecPath = (OS: string, outputPath: string) => {
+
+        let fileName = {
+            'win': 'firefox.exe',
+            'linux': 'firefox',
+        }[OS]
+
+        if (OS === 'osx') {
+            fileName = fs.readdirSync(outputPath).find(f => f.endsWith('.app'))
+        }
+
+        if (!fileName) {
+            return undefined
+        }
+
+        return {
+            'osx': path.join(outputPath, fileName, 'Contents', 'MacOS', 'firefox-bin'),
+        }[OS] ?? path.join(outputPath, fileName)
+    }
+
+    if (fs.existsSync(outputPath)) {
+
+        const executablePath = getExecPath(OS, outputPath)
+        const execFound = executablePath && fs.existsSync(executablePath)
+
+        !execFound && log.info(`Could not find Firefox executable in "${outputPath}"`)
+
+        if (execFound) {
+            const props = { executablePath, buildId: channel, platform }
+
+            let shouldDownload = false
+
+            if (channel === 'FIREFOX_NIGHTLY') {
+
+                // Check latest nightly, updates often
+                shouldDownload = await got.head(url)
+                    .then(async res => {
+
+                        const header = res.headers['last-modified']
+                        const lastModified: number = new Date(header ?? NaN).valueOf()
+
+                        return (!!header && !Number.isNaN(lastModified)) && await fsp
+                            .lstat(executablePath)
+                            .then(exec => exec.birthtimeMs < lastModified)
+                    })
+                    .catch(err => { log.error(err); return false })
+            }
+
+            if (!shouldDownload){
+                return props
+            }
+        }
+    }
+
+    await downloadFile(url, archivePath, cbProgressFirefox)
+
+    await unpackArchive(archivePath, outputPath)
+
+    if (fs.existsSync(archivePath)) {
+        await unlink(archivePath).catch(err => log.error(err))
+    }
+
+    const executablePath = getExecPath(OS, outputPath)
+
+    return { executablePath, buildId: channel, platform: arch }
+}
```

Comment:
maybe we should move all browser related code into a `chrome.ts` or `firefox.ts`

Reply:
done